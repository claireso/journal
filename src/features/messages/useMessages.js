import React, { useState, useContext } from 'react'
import PropTypes from 'prop-types'

const MessagesContext = React.createContext()

const useMessagesContext = () => {
  const context = useContext(MessagesContext)

  if (context === undefined) {
    throw new Error('useMessagesContext must be used within a MessagesProvider')
  }

  return context
}

export const MessagesProvider = ({ children }) => {
  const [state, setState] = useState([])

  return <MessagesContext.Provider value={[state, setState]}>{children}</MessagesContext.Provider>
}

MessagesProvider.propTypes = {
  children: PropTypes.any
}

const useMessages = () => {
  const [messages, setMessages] = useMessagesContext()

  const closeMessage = (index) => {
    setMessages([...messages.slice(0, index), ...messages.slice(index + 1)])
  }

  const closeAllMessages = () => setMessages([])

  const displayMessage = (status) => (message) => {
    let index = messages.length

    if (message.key) {
      const newIndex = messages.findIndex((_message) => _message.key === message.key)

      if (newIndex > -1) {
        index = newIndex
      }
    }

    setMessages([
      ...messages.slice(0, index),
      {
        status,
        ...message
      },
      ...messages.slice(index + 1)
    ])
  }

  const displaySuccessMessage = displayMessage('success')
  const displayErrorMessage = displayMessage('error')

  return [
    messages,
    {
      displaySuccessMessage,
      displayErrorMessage,
      closeMessage,
      closeAllMessages
    }
  ]
}

export default useMessages
