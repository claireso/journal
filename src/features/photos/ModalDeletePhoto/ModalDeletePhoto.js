import PropTypes from 'prop-types'

import { PrimaryButton, SecondaryButton } from '@components/Buttons'
import { Heading1 } from '@components/Headings'
import Text from '@components/Text'

import usePhotos from '@features/photos/usePhotos'

const ModalDeletePhoto = ({ onClose, id }) => {
  const [{ isProcessing }, { deletePhoto }] = usePhotos()

  const onCancel = () => onClose()

  const onConfirm = async () => {
    await deletePhoto(id)
    onClose()
  }

  return (
    <>
      <Heading1>Are you sure?</Heading1>
      <p>This action is irreversible</p>
      <Text align="right">
        <SecondaryButton onClick={onCancel}> Cancel </SecondaryButton>
        <PrimaryButton onClick={onConfirm} isLoading={isProcessing}>
          {' '}
          Yes{' '}
        </PrimaryButton>
      </Text>
    </>
  )
}

ModalDeletePhoto.propTypes = {
  onClose: PropTypes.func,
  id: PropTypes.number.isRequired
}

ModalDeletePhoto.defaultProps = {
  onClose: () => {}
}

export default ModalDeletePhoto
