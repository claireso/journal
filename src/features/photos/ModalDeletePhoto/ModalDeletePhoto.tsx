import { useCallback } from 'react'

import { Photo } from '@models'
import { ButtonPrimary, ButtonSecondary } from '@components/Buttons'
import { Heading1 } from '@components/Headings'
import Text from '@components/Text'

interface ModalDeletePhoto {
  id: Photo['id']
  onCancel: () => void
  onConfirm: (id: number) => void
  isProcessing?: boolean
}

const ModalDeletePhoto = ({ onCancel, onConfirm, id, isProcessing = false }: ModalDeletePhoto) => {
  const onClickConfirm = useCallback(() => {
    onConfirm(id)
  }, [id, onConfirm])

  return (
    <>
      <Heading1>Are you sure?</Heading1>
      <p>This action is irreversible</p>
      <Text align="right">
        <ButtonSecondary onClick={onCancel}> Cancel </ButtonSecondary>
        <ButtonPrimary onClick={onClickConfirm} isLoading={isProcessing}>
          {' '}
          Yes{' '}
        </ButtonPrimary>
      </Text>
    </>
  )
}

export default ModalDeletePhoto
