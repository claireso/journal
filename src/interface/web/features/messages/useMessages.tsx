'use client'

import React, { useState, useContext } from 'react'
import PropTypes from 'prop-types'

interface Message {
  key?: string
  status: 'default' | 'success' | 'error' | 'info'
  message: string
}

interface Actions {
  displaySuccessMessage: (data: Omit<Message, 'status'>) => void
  displayErrorMessage: (data: Omit<Message, 'status'>) => void
  closeMessage: (index: number) => void
  closeAllMessages: () => void
}

const INITIAL_STATE: Message[] = []

type TMessagesContext<T> = [T, React.Dispatch<React.SetStateAction<T>>]

const MessagesContext = React.createContext<TMessagesContext<Message[]>>([null!, null!])

interface MessagesProviderProps {
  children: React.ReactNode
}

export const MessagesProvider = ({ children }: MessagesProviderProps) => {
  const [state, setState] = useState(INITIAL_STATE)

  return <MessagesContext.Provider value={[state, setState]}>{children}</MessagesContext.Provider>
}

const useMessagesContext = () => {
  const context = useContext(MessagesContext)

  if (context === undefined) {
    throw new Error('useMessagesContext must be used within a MessagesProvider')
  }

  return context
}

MessagesProvider.propTypes = {
  children: PropTypes.any
}

const useMessages = (): [Message[], Actions] => {
  const [messages, setMessages] = useMessagesContext()

  const closeMessage = (index: number) => {
    setMessages([...messages.slice(0, index), ...messages.slice(index + 1)])
  }

  const closeAllMessages = () => setMessages([])

  const displayMessage = (status: Message['status']) => (message: Omit<Message, 'status'>) => {
    let index = messages.length

    if (message.key) {
      const newIndex = messages.findIndex((_message: Message) => _message.key === message.key)

      if (newIndex > -1) {
        index = newIndex
      }
    }

    setMessages([
      ...messages.slice(0, index),
      {
        status,
        ...message
      },
      ...messages.slice(index + 1)
    ])
  }

  const displaySuccessMessage = displayMessage('success')
  const displayErrorMessage = displayMessage('error')

  return [
    messages,
    {
      displaySuccessMessage,
      displayErrorMessage,
      closeMessage,
      closeAllMessages
    }
  ]
}

export default useMessages
