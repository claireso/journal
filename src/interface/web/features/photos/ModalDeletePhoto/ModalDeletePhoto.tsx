import { useCallback } from 'react'

import type { PhotoDto } from '@dto'
import { ButtonPrimary, ButtonSecondary } from '@web/components/Buttons'
import { Heading1 } from '@web/components/Headings'
import Text from '@web/components/Text'

interface ModalDeletePhoto {
  id: PhotoDto['id']
  onCancel: () => void
  onConfirm: (id: number) => void
  isProcessing?: boolean
}

const ModalDeletePhoto = ({ onCancel, onConfirm, id, isProcessing = false }: ModalDeletePhoto) => {
  const onClickConfirm = useCallback(() => {
    onConfirm(id)
  }, [id, onConfirm])

  return (
    <>
      <Heading1>Are you sure?</Heading1>
      <p>This action is irreversible</p>
      <Text align="right">
        <ButtonSecondary onClick={onCancel}> Cancel </ButtonSecondary>
        <ButtonPrimary onClick={onClickConfirm} isLoading={isProcessing}>
          {' '}
          Yes{' '}
        </ButtonPrimary>
      </Text>
    </>
  )
}

export default ModalDeletePhoto
