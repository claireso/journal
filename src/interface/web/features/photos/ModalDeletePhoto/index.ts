import ModalDeletePhoto from './ModalDeletePhoto'

export default ModalDeletePhoto
