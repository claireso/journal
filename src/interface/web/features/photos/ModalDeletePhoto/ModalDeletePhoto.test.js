import { fireEvent, render, screen } from '@testing-library/react'

import { MessagesProvider } from '../../messages/useMessages'
import ModalDeletePhoto from './ModalDeletePhoto'

describe('<ModalDeletePhoto />', () => {
  const renderComponent = (props) =>
    render(
      <MessagesProvider>
        <ModalDeletePhoto id={188} {...props} />
      </MessagesProvider>
    )

  it('should render component', () => {
    const { asFragment } = renderComponent()

    expect(asFragment()).toMatchSnapshot()
  })

  it('should not delete subscription', () => {
    const props = {
      onCancel: jest.fn()
    }

    renderComponent(props)

    fireEvent.click(screen.getByText('Cancel'))

    expect(props.onCancel).toHaveBeenCalled()
  })

  it('should delete subscription', async () => {
    const props = {
      onConfirm: jest.fn()
    }

    renderComponent(props)

    fireEvent.click(screen.getByText('Yes'))

    expect(props.onConfirm).toHaveBeenCalledWith(188)
  })
})
