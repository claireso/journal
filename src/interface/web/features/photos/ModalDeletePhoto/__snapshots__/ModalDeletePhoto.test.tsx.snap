// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ModalDeletePhoto /> should render component 1`] = `
<DocumentFragment>
  <p>
    This canâ€™t be undone
  </p>
  <div
    class="styles_layerStyle__ulchk90"
  >
    <button
      class="styles__1mdc28jf styles_layerStyle__1mdc28j0 styles_button_variant_neutral__1mdc28jg styles_layerStyle__1mdc28j1 styles_button_size_lg__1mdc28jm styles_layerStyle__1mdc28j7"
    >
      Cancel
    </button>
    <button
      class="styles__1mdc28jf styles_layerStyle__1mdc28j0 styles_button_variant_danger__1mdc28ji styles_layerStyle__1mdc28j3 styles_button_size_lg__1mdc28jm styles_layerStyle__1mdc28j7"
    >
      Delete
    </button>
  </div>
</DocumentFragment>
`;
